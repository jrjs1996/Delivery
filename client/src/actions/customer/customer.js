import axios from 'axios';
import {
  ADD_CUSTOMER,
  DELETE_CURRENT_CUSTOMER,
  DELETE_CUSTOMER,
  ERROR,
  FETCH_CUSTOMERS,
  UPDATE_CURRENT_CUSTOMER,
  UPDATE_CUSTOMER,
} from '../types';
import { saveToken, setAuthHeader, removeToken } from '../../utils/token';

export const addCustomer = postData => async (dispatch) => {
  try {
    const res = await axios.post('/api/customers/', postData);
    dispatch({
      type: ADD_CUSTOMER,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to add customer',
    });
  }
};

export const deleteCustomer = customerId => async (dispatch) => {
  try {
    await axios.delete(`/api/customers/${customerId}`, setAuthHeader());
    dispatch({
      type: DELETE_CUSTOMER,
      payload: { _id: customerId },
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to delete customer',
    });
  }
};

export const fetchCustomers = () => async (dispatch) => {
  try {
    const res = await axios.get('/api/customers/', setAuthHeader());
    dispatch({
      type: FETCH_CUSTOMERS,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to fetch customers.',
    });
  }
};

export const getCurrentCustomerInfo = () => async (dispatch) => {
  try {
    const res = await axios.get('/api/customers/login/', setAuthHeader());
    dispatch({
      type: UPDATE_CURRENT_CUSTOMER,
      payload: res.data,
    });
  } catch (error) {
    if (error.response.status === 401) {
      console.log('h1h1');
      removeToken();
      return;
    }
    console.log('h1h1');
    throw error;
  }
};

export const getCustomer = userId => (dispatch) => {
  axios.get('/api/customers/info/', setAuthHeader()).then((res) => {
    dispatch({
      type: UPDATE_CURRENT_CUSTOMER,
      payload: res.data,
    });
  });
};

export const login = postData => async (dispatch) => {
  try {
    const res = await axios.post('/api/customers/login/', postData);
    saveToken(res.data.token, false);
    dispatch({
      type: UPDATE_CURRENT_CUSTOMER,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: error.toString(),
    });
  }
};

export const logout = () => (dispatch) => {
  try {
    removeToken();
    dispatch({
      type: DELETE_CURRENT_CUSTOMER,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to log out.',
    });
  }
};

export const updateCustomer = putData => async (dispatch) => {
  try {
    await axios.put(`/api/customers/${putData._id}`, putData, setAuthHeader());
    const payload = putData;
    dispatch({
      type: UPDATE_CUSTOMER,
      payload,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to update customer',
    });
  }
};
