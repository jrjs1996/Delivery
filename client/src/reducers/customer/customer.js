import {
  ADD_CUSTOMER,
  DELETE_CUSTOMER,
  FETCH_CUSTOMERS,
  UPDATE_CURRENT_CUSTOMER,
  UPDATE_CUSTOMER,
} from '../../actions/types';
import {
  updateItem, deleteItem, insertItem, createItems,
} from '../utils/utils';

const initialState = {
  items: {},
  currentCustomer: {},
};

export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_CUSTOMER:
      return {
        ...state,
        items: insertItem(state.items, action.payload, '_id'),
      };
    case DELETE_CUSTOMER:
      return {
        ...state,
        items: deleteItem(state.items, action.payload, '_id'),
      };
    case FETCH_CUSTOMERS:
      return {
        ...state,
        items: createItems(action.payload, '_id'),
      };
    case UPDATE_CURRENT_CUSTOMER:
      return {
        ...state,
        currentCustomer: action.payload,
      };
    case UPDATE_CUSTOMER:
      return {
        ...state,
        items: updateItem(state.items, action.payload, '_id'),
      };
    default:
      return state;
  }
}
