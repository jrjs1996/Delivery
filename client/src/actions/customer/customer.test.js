import configureStore from 'redux-mock-store';
import * as customerActions from './customer';
import mockAxios from '../../../__mocks__/axios';

const mockStore = configureStore();
const store = mockStore();

describe('Add customer', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches ADD_CUSTOMER with response data', async () => {
    await customerActions.addCustomer({ mockInfo: 'MockInfo' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches error on error', async () => {
    mockAxios.giveError(mockAxios.post);
    await customerActions.addCustomer({ mockInfo: 'MockInfo' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Delete customer', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches DELETE_CUSTOMER with customer id', async () => {
    await customerActions.deleteCustomer('Mock id')(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches error on error', async () => {
    mockAxios.giveError(mockAxios.delete);
    await customerActions.deleteCustomer({ mockInfo: 'MockInfo' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Fetch customers', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches FETCH_CUSTOMERS with response data', async () => {
    await customerActions.fetchCustomers()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches error on error', async () => {
    mockAxios.giveError(mockAxios.get);
    await customerActions.fetchCustomers()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Update customer', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches UPDATE_CUSTOMER with given data', async () => {
    await customerActions.updateCustomer({ mockData: 'MockData' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches error on error', async () => {
    mockAxios.giveError(mockAxios.put);
    await customerActions.updateCustomer()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});
