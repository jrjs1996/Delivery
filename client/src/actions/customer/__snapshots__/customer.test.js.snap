// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add customer Dispatches ADD_CUSTOMER with response data 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
    },
    "type": "ADD_CUSTOMER",
  },
]
`;

exports[`Add customer Dispatches error on error 1`] = `
Array [
  Object {
    "payload": "Unable to add customer",
    "type": "ERROR",
  },
]
`;

exports[`Delete customer Dispatches DELETE_CUSTOMER with customer id 1`] = `
Array [
  Object {
    "payload": Object {
      "_id": "Mock id",
    },
    "type": "DELETE_CUSTOMER",
  },
]
`;

exports[`Delete customer Dispatches error on error 1`] = `
Array [
  Object {
    "payload": "Unable to delete customer",
    "type": "ERROR",
  },
]
`;

exports[`Fetch customers Dispatches FETCH_CUSTOMERS with response data 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
    },
    "type": "FETCH_CUSTOMERS",
  },
]
`;

exports[`Fetch customers Dispatches error on error 1`] = `
Array [
  Object {
    "payload": "Unable to fetch customers.",
    "type": "ERROR",
  },
]
`;

exports[`Login Dispatches UPDATE_CURRENT_CUSTOMER with response data 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
      "token": "MockToken",
    },
    "type": "UPDATE_CURRENT_CUSTOMER",
  },
]
`;

exports[`Update customer Dispatches UPDATE_CUSTOMER with given data 1`] = `
Array [
  Object {
    "payload": Object {
      "mockData": "MockData",
    },
    "type": "UPDATE_CUSTOMER",
  },
]
`;

exports[`Update customer Dispatches error on error 1`] = `
Array [
  Object {
    "payload": "Unable to update customer",
    "type": "ERROR",
  },
]
`;
